### PR TITLE
bpo-36146: Refactor setup.py

### PR DESCRIPTION
**Pull request by vstinner**
The detect_modules() method of setup.py became longer and longer over
the years. It is now 1128 lines long. It's way too long: it becomes
very hard to track the lifetime of a variable and many variables are
overriden on purpose or not. Shorter functions help to track the
lifetime of variables, ease review and reduce the number of bugs.

Changes:

* Fix detect_ctypes(): it no longer modifies global inc_dirs on
  macOS
* self.srcdir is now always an absolute path
* Add MACOS and MS_WINDOWS constants
* Add TEST_EXTENSIONS to allow to not build extensions used to test
  Python (ex: _testcapi)
* Add MATH_LIBS which is the libm library

Other refactoring changes:

* Reorder imports
* Replace "from distutils.errors import *"
  with "from distutils.errors import CCompilerError, DistutilsError"
  to be able to use static analyzers like pyflakes
* Rename globals to upper case to better distinguish if a variable is
  global or locale:

  * Rename cross_compiling to CROSS_COMPILING
  * Rename host_platform to HOST_PLATFORM
  * Rename disabled_module_list to DISABLED_MODULE_LIST

* Don't run code at module top level, but at the beginning of main()
* Move 'missing' and 'srcdir' from detect_modules() local variable to
  PyBuildExt attributes
* Remove exts local variable from detect_modules(): add add_ext()
  method which append directly the module to self.extensions.
* _detect_openssl() and _detect_nis() now directly add the extension
  and have no return value.
* Split very long detect_modules() method into subfunctions, add
  many "detect_xxx" methods.

https://bugs.python.org/issue36146

### CODE REVIEW

**Comment by vstinner** [comment]
@pablogsal @serhiy-storchaka @methane @gpshead @pxinwr: Would you mind to review this large refactoring change? I basically moved a lot of code, but I made some additional changes described in the commit message (and PR description).

**Comment by vstinner** [comment]
Since the change shouldn't modify the behavior of setup.py of any way, I decided to skip the bpo number and NEWS entry.

**Comment by matrixise** [comment]
@vstinner don't you prefer to use `pathlib.Path` instead of `os.path.XYZ` ?

File: setup.py [hunk 1]
```diff
@@ -220,23 +263,37 @@ def find_module_file(module, dirlist):
...
+        self.inc_dirs = None
+        self.lib_dirs = None
         if '-j' in os.environ.get('MAKEFLAGS', ''):
             self.parallel = True
 
+    def add_ext(self, ext):
+        self.extensions.append(ext)
+
     def build_extensions(self):
+        self.srcdir = sysconfig.get_config_var('srcdir')
```
Comment (matrixise) [inline setup.py:286]: maybe you could use `pathlib.Path` ?

File: setup.py [hunk 2]
```diff
@@ -425,7 +477,7 @@ def check_extension_import(self, ext):
 
         # Workaround for Cygwin: Cygwin currently has fork issues when many
         # modules have been imported
-        if host_platform == 'cygwin':
+        if HOST_PLATFORM == 'cygwin':
```
Comment (matrixise) [inline setup.py]: add `CYGWIN = (HOST_PLATFORM == 'cygwin')` after `MS_WINDOWS` ?

**Comment by matrixise** [comment]
@vstinner  maybe you prefer to keep the compatibility with `python2` and avoid `pathlib.Path`.

File: setup.py (hunk 1 above)
Comment (vstinner) [reply setup.py:286]: setup.py is used to build Python. When setup.py is called, most Python modules are not available. I prefer to use the bare minimum from the stdlib. For example, "import subprocess" fails with "ModuleNotFoundError: No module named _posixsubprocess". That's why os.system() is used.

Moreover, I prefer to limit changes in this PR.

File: setup.py (hunk 2 above)
Comment (vstinner) [reply setup.py]: Yeah, I hesitated to add this one. Done, I added CYGWIN.

**Comment by matrixise** [comment]
@vstinner excepted the issues with ftp://www.pythontest.net your code could be merged.

**Comment by ned-deily** [comment]
Yikes!  Please do not merge this until I've had a chance to better review and test various build configurations.  It would be a lot easier without so many gratuitous changes.

**Comment by vstinner** [comment]
> Yikes! Please do not merge this until I've had a chance to better review and test various build configurations. It would be a lot easier without so many gratuitous changes.

I'm sorry, my intent wasn't to scare you :-) I agree that this PR modifies too many unrelated things at once, and so I close it in favor of shorter changes which would be easier to review.

I created https://bugs.python.org/issue36146 to track this work and I started with PR #12093 to convert globals to upper case and add some new constants. This PR should be straightforward not controversial in any way :-)

**Pull request closed without merging**